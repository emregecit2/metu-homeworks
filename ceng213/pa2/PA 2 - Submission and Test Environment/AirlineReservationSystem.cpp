#include "AirlineReservationSystem.h"

void AirlineReservationSystem::addPassenger(const std::string &firstname, const std::string &lastname) {
    passengers.insert(Passenger(firstname, lastname));
}

Passenger *AirlineReservationSystem::searchPassenger(const std::string &firstname, const std::string &lastname) {
    return &passengers.search(Passenger(firstname, lastname))->data;
}

void AirlineReservationSystem::addFlight(const std::string &flightCode, const std::string &departureTime, const std::string &arrivalTime, const std::string &departureCity, const std::string &arrivalCity, int economyCapacity, int businessCapacity) {
    flights.insert(Flight(flightCode, departureTime, arrivalTime, departureCity, arrivalCity, economyCapacity, businessCapacity));
}

std::vector<Flight *> AirlineReservationSystem::searchFlight(const std::string &departureCity, const std::string &arrivalCity) {
    std::vector<Flight *> v;
    for(BSTNode<Flight> *curr = flights.getRoot(); curr; curr=flights.getSuccessor(curr, preorder)){
        if(curr->data.getDepartureCity()==departureCity && curr->data.getArrivalCity()==arrivalCity){
            v.push_back(&curr->data);
        }
    }
    return v;
}

void AirlineReservationSystem::issueTicket(const std::string &firstname, const std::string &lastname, const std::string &flightCode, TicketType ticketType) {
    for(BSTNode<Flight> *curr = flights.getRoot(); curr; curr=flights.getSuccessor(curr, preorder)){
        if(curr->data.getFlightCode()==flightCode){
            curr->data.addTicket(Ticket(searchPassenger(firstname, lastname), &curr->data, ticketType));
            break;
        }
    }
}


void AirlineReservationSystem::saveFreeTicketRequest(const std::string &firstname, const std::string &lastname, const std::string &flightCode, TicketType ticketType) {
    for(BSTNode<Flight> *curr = flights.getRoot(); curr; curr=flights.getSuccessor(curr, preorder)){
        if(curr->data.getFlightCode()==flightCode){
            freeTicketRequests.enqueue(Ticket(searchPassenger(firstname, lastname), &curr->data, ticketType));
            break;
        }
    }
}

void AirlineReservationSystem::executeTheFlight(const std::string &flightCode) {
    int len = freeTicketRequests.size();
    for(BSTNode<Flight> *curr = flights.getRoot(); curr; curr=flights.getSuccessor(curr, preorder)){
        if(curr->data.getFlightCode()==flightCode){
            Flight &flight = curr->data;
            flight.setCompleted(true);
            for(int i=0; i<len; i++){
                Ticket ticket = freeTicketRequests.dequeue();
                if(!(ticket.getFlight()->getFlightCode()==flightCode && flight.addTicket(ticket))){
                    freeTicketRequests.enqueue(ticket);
                }
            }
            break;
        }
    }
}

void AirlineReservationSystem::print() const {
    std::cout << "# Printing the airline reservation system ..." << std::endl;

    std::cout << "# Passengers:" << std::endl;
    passengers.print(inorder);

    std::cout << "# Flights:" << std::endl;
    flights.print(inorder);

    std::cout << "# Free ticket requests:" << std::endl;
    freeTicketRequests.print();

    std::cout << "# Printing is done." << std::endl;
}