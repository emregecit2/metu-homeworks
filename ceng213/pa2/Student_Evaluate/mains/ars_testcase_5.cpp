#include <iostream>

#include "AirlineReservationSystem.h"
//#include "../AirlineReservationSystem.h"

/*
 * Case 5 : Default constructor; addPassenger, addFlight, searchFlight; print.
 */
int main() {

    std::cout << "-> Creating an airline reservation system instance with default constructor." << std::endl;

    AirlineReservationSystem ars;

    std::cout << "-> Making some passenger insertions to the system." << std::endl;

    ars.addPassenger("jack", "sophia");
    ars.addPassenger("archie", "rosie");
    ars.addPassenger("harry", "isabella");
    ars.addPassenger("oscar", "lily");
    ars.addPassenger("leo", "ivy");
    ars.addPassenger("kate", "sophia");

    std::cout << "-> Making some flight insertions to the system." << std::endl;

    ars.addFlight("DS7050", "20/12/2021 10:00", "20/12/2021 11:30", "Ankara", "Istanbul", 200, 10);
    ars.addFlight("DS7030", "20/12/2021 11:00", "20/12/2021 12:30", "Ankara", "Brussels", 200, 10);
    ars.addFlight("DS7070", "20/12/2021 12:00", "20/12/2021 13:30", "Ankara", "Amsterdam", 200, 10);
    ars.addFlight("DS7040", "20/12/2021 13:00", "20/12/2021 14:30", "Ankara", "London", 200, 10);
    ars.addFlight("DS7060", "20/12/2021 14:00", "20/12/2021 15:30", "Ankara", "Rome", 200, 10);
    ars.addFlight("DS7020", "20/12/2021 15:00", "20/12/2021 16:30", "Ankara", "London", 100, 5);
    ars.addFlight("DS7080", "20/12/2021 16:00", "20/12/2021 17:30", "Ankara", "London", 50, 0);

    std::cout << "-> Printing the system." << std::endl;

    ars.print();

    std::string departureCitiesToSearch[] = {"Ankara", "Ankara"};
    std::string arrivalCitiesToSearch[] = {"Istanbul", "London"};

    for (int i = 0 ; i < 2 ; ++i) {
        std::cout << "-> Searching flights from " << departureCitiesToSearch[i] << " to " << arrivalCitiesToSearch[i] << "." << std::endl;
        std::vector<Flight *> flights = ars.searchFlight(departureCitiesToSearch[i], arrivalCitiesToSearch[i]);
        if (flights.empty()) {
            std::cout << "no flights from " << departureCitiesToSearch[i] << " to " << arrivalCitiesToSearch[i] << " exists." << std::endl;
        } else {
            for (int j = 0 ; j < flights.size() ; ++j) {
                std::cout << *flights[j] << std::endl;
            }
        }
    }

    std::cout << "-> Printing the system." << std::endl;

    ars.print();

    return 0;
}
