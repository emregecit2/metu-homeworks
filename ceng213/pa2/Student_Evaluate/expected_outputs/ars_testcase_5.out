-> Creating an airline reservation system instance with default constructor.
-> Making some passenger insertions to the system.
-> Making some flight insertions to the system.
-> Printing the system.
# Printing the airline reservation system ...
# Passengers:
BST_inorder{
	Passenger(firstname: harry, lastname: isabella),
	Passenger(firstname: leo, lastname: ivy),
	Passenger(firstname: oscar, lastname: lily),
	Passenger(firstname: archie, lastname: rosie),
	Passenger(firstname: jack, lastname: sophia),
	Passenger(firstname: kate, lastname: sophia)
}
# Flights:
BST_inorder{
	Flight(flightCode: DS7020, departureTime: 20/12/2021 15:00, arrivalTime: 20/12/2021 16:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 100, businessCapacity: 5, completed: no, tickets: []),
	Flight(flightCode: DS7030, departureTime: 20/12/2021 11:00, arrivalTime: 20/12/2021 12:30, departureCity: Ankara, arrivalCity: Brussels, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7040, departureTime: 20/12/2021 13:00, arrivalTime: 20/12/2021 14:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7050, departureTime: 20/12/2021 10:00, arrivalTime: 20/12/2021 11:30, departureCity: Ankara, arrivalCity: Istanbul, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7060, departureTime: 20/12/2021 14:00, arrivalTime: 20/12/2021 15:30, departureCity: Ankara, arrivalCity: Rome, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7070, departureTime: 20/12/2021 12:00, arrivalTime: 20/12/2021 13:30, departureCity: Ankara, arrivalCity: Amsterdam, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7080, departureTime: 20/12/2021 16:00, arrivalTime: 20/12/2021 17:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 50, businessCapacity: 0, completed: no, tickets: [])
}
# Free ticket requests:
Queue{}
# Printing is done.
-> Searching flights from Ankara to Istanbul.
Flight(flightCode: DS7050, departureTime: 20/12/2021 10:00, arrivalTime: 20/12/2021 11:30, departureCity: Ankara, arrivalCity: Istanbul, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: [])
-> Searching flights from Ankara to London.
Flight(flightCode: DS7020, departureTime: 20/12/2021 15:00, arrivalTime: 20/12/2021 16:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 100, businessCapacity: 5, completed: no, tickets: [])
Flight(flightCode: DS7040, departureTime: 20/12/2021 13:00, arrivalTime: 20/12/2021 14:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: [])
Flight(flightCode: DS7080, departureTime: 20/12/2021 16:00, arrivalTime: 20/12/2021 17:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 50, businessCapacity: 0, completed: no, tickets: [])
-> Printing the system.
# Printing the airline reservation system ...
# Passengers:
BST_inorder{
	Passenger(firstname: harry, lastname: isabella),
	Passenger(firstname: leo, lastname: ivy),
	Passenger(firstname: oscar, lastname: lily),
	Passenger(firstname: archie, lastname: rosie),
	Passenger(firstname: jack, lastname: sophia),
	Passenger(firstname: kate, lastname: sophia)
}
# Flights:
BST_inorder{
	Flight(flightCode: DS7020, departureTime: 20/12/2021 15:00, arrivalTime: 20/12/2021 16:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 100, businessCapacity: 5, completed: no, tickets: []),
	Flight(flightCode: DS7030, departureTime: 20/12/2021 11:00, arrivalTime: 20/12/2021 12:30, departureCity: Ankara, arrivalCity: Brussels, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7040, departureTime: 20/12/2021 13:00, arrivalTime: 20/12/2021 14:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7050, departureTime: 20/12/2021 10:00, arrivalTime: 20/12/2021 11:30, departureCity: Ankara, arrivalCity: Istanbul, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7060, departureTime: 20/12/2021 14:00, arrivalTime: 20/12/2021 15:30, departureCity: Ankara, arrivalCity: Rome, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7070, departureTime: 20/12/2021 12:00, arrivalTime: 20/12/2021 13:30, departureCity: Ankara, arrivalCity: Amsterdam, economyCapacity: 200, businessCapacity: 10, completed: no, tickets: []),
	Flight(flightCode: DS7080, departureTime: 20/12/2021 16:00, arrivalTime: 20/12/2021 17:30, departureCity: Ankara, arrivalCity: London, economyCapacity: 50, businessCapacity: 0, completed: no, tickets: [])
}
# Free ticket requests:
Queue{}
# Printing is done.
